src/compiler/factory/node_factory.rs

```diff
@@ -35,7 +39,7 @@ bitflags! {
     }
 }
 
-impl NodeFactory {
+impl<TBaseNodeFactory: 'static + BaseNodeFactory> NodeFactory<TBaseNodeFactory> {
```

Comment:
In order to have `ParenthesizerRules` be object-safe I couldn't have `<TBaseNodeFactory: BaseNodeFactory>` generics in the trait method signatures, so refactored so that `<TBaseNodeFactory>` is on the `NodeFactory`/`ParenthesizerRules` rather than on each individual method call